File: webrtc-jni/src/main/cpp/src/api/RTCRtpCodecParameters.cpp
/*
 * Copyright 2019 Alex Andres
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "api/RTCRtpCodecParameters.h"
#include "JavaEnums.h"
#include "JavaHashMap.h"
#include "JavaObject.h"
#include "JavaPrimitive.h"
#include "JavaRef.h"
#include "JavaString.h"
#include "JavaUtils.h"
#include "JNI_WebRTC.h"

namespace jni
{
	namespace RTCRtpCodecParameters
	{
		JavaLocalRef<jobject> toJava(JNIEnv * env, const webrtc::RtpCodecParameters & parameters)
		{
			jint payloadType = static_cast<jint>(parameters.payload_type);
			JavaLocalRef<jobject> mediaType = JavaEnums::toJava(env, parameters.kind);
			JavaLocalRef<jstring> codecName = JavaString::toJava(env, parameters.name);
			JavaLocalRef<jobject> clockRate = nullptr;
			JavaLocalRef<jobject> channels = nullptr;
			JavaHashMap paramMap(env);

			for (const std::pair<std::string, std::string> & param : parameters.parameters) {
				JavaLocalRef<jstring> key = JavaString::toJava(env, param.first);
				JavaLocalRef<jstring> value = JavaString::toJava(env, param.second);

				paramMap.put(key, value);
			}

			if (parameters.clock_rate.has_value()) {
				clockRate = Integer::create(env, parameters.clock_rate.value());
			}
			if (parameters.num_channels.has_value()) {
				channels = Integer::create(env, parameters.num_channels.value());
			}

			const auto javaClass = JavaClasses::get<JavaRTCRtpCodecParametersClass>(env);

			jobject object = env->NewObject(javaClass->cls, javaClass->ctor, payloadType, mediaType.get(), codecName.get(), clockRate.get(), channels.get(), ((JavaLocalRef<jobject>)paramMap).get());
			ExceptionCheck(env);

			return JavaLocalRef<jobject>(env, object);
		}

		webrtc::RtpCodecParameters toNative(JNIEnv * env, const JavaRef<jobject> & parameters)
		{
			const auto javaClass = JavaClasses::get<JavaRTCRtpCodecParametersClass>(env);

			JavaObject obj(env, parameters);

			webrtc::RtpCodecParameters params;
			params.payload_type = obj.getInt<int>(javaClass->payloadType);
			params.kind = JavaEnums::toNative<cricket::MediaType>(env, obj.getObject(javaClass->mediaType));
			params.name = JavaString::toNative(env, obj.getString(javaClass->codecName));

			for (const auto & entry : JavaHashMap(env, obj.getObject(javaClass->parameters))) {
				std::string key = JavaString::toNative(env, static_java_ref_cast<jstring>(env, entry.first));
				std::string value = JavaString::toNative(env, static_java_ref_cast<jstring>(env, entry.second));

				params.parameters.emplace(key, value);
			}

			auto clockRate = obj.getObject(javaClass->clockRate);
			auto channels = obj.getObject(javaClass->channels);

			if (clockRate.get()) {
				params.clock_rate = Integer::getValue(env, clockRate);
			}
			if (channels.get()) {
				params.num_channels = Integer::getValue(env, channels);
			}

			return params;
		}

		JavaRTCRtpCodecParametersClass::JavaRTCRtpCodecParametersClass(JNIEnv * env)
		{
			cls = FindClass(env, PKG"RTCRtpCodecParameters");

			ctor = GetMethod(env, cls, "<init>", "(IL" PKG_MEDIA "MediaType;" STRING_SIG INTEGER_SIG INTEGER_SIG MAP_SIG ")V");

			payloadType = GetFieldID(env, cls, "payloadType", "I");
			mediaType = GetFieldID(env, cls, "mediaType", "L" PKG_MEDIA "MediaType;");
			codecName = GetFieldID(env, cls, "codecName", STRING_SIG);
			clockRate = GetFieldID(env, cls, "clockRate", INTEGER_SIG);
			channels = GetFieldID(env, cls, "channels", INTEGER_SIG);
			parameters = GetFieldID(env, cls, "parameters", MAP_SIG);
		}
	}
}

File: webrtc/src/main/java/dev/onvoid/webrtc/RTCRtpCodecParameters.java
/*
 * Copyright 2019 Alex Andres
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package dev.onvoid.webrtc;

import dev.onvoid.webrtc.media.MediaType;

import java.util.Map;

/**
 * Describes the configuration parameters for a single media codec.
 *
 * @author Alex Andres
 */
public class RTCRtpCodecParameters {

	/**
	 * The RTP payload type used to identify this codec.
	 */
	public final int payloadType;

	/**
	 * The media type of the codec. Equivalent to the MIME top-level type.
	 */
	public final MediaType mediaType;

	/**
	 * Used to identify the codec. Equivalent to the MIME subtype.
	 */
	public final String codecName;

	/**
	 * The codec clock rate expressed in Hertz.
	 */
	public final Integer clockRate;

	/**
	 * When present, indicates the number of audio channels (mono=1, stereo=2).
	 */
	public final Integer channels;

	/**
	 * The "format specific parameters" field from the "a=fmtp" line in the SDP
	 * corresponding to the codec, if one exists. For an RTCRtpSender, these
	 * parameters come from the remote description, and for an RTCRtpReceiver,
	 * they come from the local description.
	 */
	public final Map<String, String> parameters;


	/**
	 * Creates an instance of RTCRtpCodecParameters with the specified media
	 * codec parameters.
	 *
	 * @param payloadType The RTP payload type used to identify the codec.
	 * @param mediaType   The media type of the codec.
	 * @param codecName   The codec name. Equivalent to the MIME subtype.
	 * @param clockRate   The codec clock rate expressed in Hertz.
	 * @param channels    The number of audio channels. Unused for video codecs.
	 * @param parameters  The "a=fmtp" parameters in the SDP.
	 */
	public RTCRtpCodecParameters(int payloadType, MediaType mediaType,
			String codecName, Integer clockRate, Integer channels,
			Map<String, String> parameters) {
		this.payloadType = payloadType;
		this.mediaType = mediaType;
		this.codecName = codecName;
		this.clockRate = clockRate;
		this.channels = channels;
		this.parameters = parameters;
	}

	@Override
	public String toString() {
		return String.format("%s [payloadType=%s, mediaType=%s, codecName=%s, clockRate=%s, channels=%s, parameters=%s]",
				RTCRtpCodecParameters.class.getSimpleName(), payloadType,
				mediaType, codecName, clockRate, channels, parameters);
	}
}
